sympy/polys/domains/field.py

```diff
@@ -68,6 +68,39 @@ def gcd(self, a, b):
 
         return self.convert(p, ring)/q
 
+    def gcdex(self, a, b):
+        """
+        Returns x, y, g such that a * x + b * y == g == gcd(a, b)
+        """
+        try:
+            ring = self.get_ring()
+            found_ring = True
+        except DomainError:
+            found_ring = False
+
+        if (not found_ring) or ring == self:
+            if found_ring:
+                d = self.one
+            else:
+                d = self.gcd(a, b)
+
+            if a == self.zero:
+                if b == self.zero:
+                    return self.zero, self.one, self.zero
+                else:
+                    return self.zero, d/b, d
+            else:
+                return d/a, self.zero, d
+
+        n1, n2 = self.numer(a), self.numer(b)
+        d1, d2 = self.denom(a), self.denom(b)
+        g = ring.gcd(d1, d2)
+        q = d1 * (d2 / g)
+        n1, n2 = n1 * (d2 / g), n2 * (d1 / g)
+
+        x, y, d = ring.gcdex(n1, n2)
+        return self.convert(x, ring)/q, self.convert(y, ring)/q, self.convert(d, ring)/q
```

Comment:
It seems like this is trying to make a fraction field behave like its ring of integers. I'm not sure whether that makes sense or not. I guess we already have things like:
```python
In [8]: QQ.gcd(QQ(1,6), QQ(1,3))
Out[8]: 1/6

In [12]: F = QQ.frac_field(x)

In [13]: F.gcd(2/F(x), F(4/x**2))
Out[13]: 2/(x**2)

In [14]: F.gcd(2*F(x), F(4*x**2))
Out[14]: 2*x
```
Is this just aiming for consistency with that sort of thing?

We should have some tests checking that `gcd` and `gcdex` are consistent across domains.

Reply:
I'm not sure if this definition of `gcdex` for fields is really what makes sense for SNF. Maybe SNF just needs the trivial definition where the gcd is always 1.

Reply:
I'm also not sure if the best way to applied SNF to fields should use field division though or if some sort of fraction-free approach would be better.

If we have
```python
A = S * M * T
```
where say the elements of `M` are in `Q[x,y]` (not a PID) we can extend to `Q(x,y)` to have a field and compute the result for A, S and T with A having all 1s. Then we will have fractions in S and T though and ideally we would want to clear those. We can clear denominators from rows or columns of a matrix e.g. to clear denominators from rows we have `S = S_d^-1 S_ff` where `S_d` is diagonal and both `S_d` and `S_ff` are fraction-free. We can also clear denominators from columns like `T = T_ff * T_d^1` in the same way. We then have
```
A = S_d^-1 * S_ff * M * T_ff * T_d^-1
S_d * A * T_d = S_ff * M * T_ff
A_ff = S_ff * M * T_ff
```
We could have a function that returns `A_ff`, `S_ff`, `T_ff`, `S_d`, `T_d` as a fraction-free representation of SNF over a field. Probably this is just coming back to the LU decomposition again though and what I am referring to here is just a fraction-free LU (and should be derived from that). We do not currently have fraction-free LU although it is not hard to make a basic implementation.